insights/lambdas/bb2-kinesis-firehose-cloudwatch-logs-processor-python.py

```diff
@@ -0,0 +1,342 @@
+            # (if the result does not have a key 'ErrorCode' OR if it does and is empty) => we do not need to re-ingest
+            if "ErrorCode" not in res or not res["ErrorCode"]:
+                continue
+
+            codes.append(res["ErrorCode"])
+            failedRecords.append(records[idx])
+
+        errMsg = "Individual error codes: " + ",".join(codes)
+
+    if len(failedRecords) > 0:
+        if attemptsMade + 1 < maxAttempts:
+            print(
+                "Some records failed while calling PutRecordBatch to Firehose stream, retrying. %s"
+                % (errMsg)
+            )
+            putRecordsToFirehoseStream(
+                streamName, failedRecords, client, attemptsMade + 1, maxAttempts
+            )
+        else:
+            raise RuntimeError(
+                "Could not put records after %s attempts. %s"
+                % (str(maxAttempts), errMsg)
+            )
+
+
+def putRecordsToKinesisStream(streamName, records, client, attemptsMade, maxAttempts):
+    failedRecords = []
+    codes = []
+    errMsg = ""
+    # if put_records throws for whatever reason, response['xx'] will error out, adding a check for a valid
+    # response will prevent this
+    response = None
+    try:
+        response = client.put_records(StreamName=streamName, Records=records)
+    except Exception as e:
+        failedRecords = records
+        errMsg = str(e)
+
+    # if there are no failedRecords (put_record_batch succeeded), iterate over the response to gather results
+    if not failedRecords and response and response["FailedRecordCount"] > 0:
+        for idx, res in enumerate(response["Records"]):
+            # (if the result does not have a key 'ErrorCode' OR if it does and is empty) => we do not need to re-ingest
+            if "ErrorCode" not in res or not res["ErrorCode"]:
+                continue
+
+            codes.append(res["ErrorCode"])
+            failedRecords.append(records[idx])
+
+        errMsg = "Individual error codes: " + ",".join(codes)
+
+    if len(failedRecords) > 0:
+        if attemptsMade + 1 < maxAttempts:
+            print(
+                "Some records failed while calling PutRecords to Kinesis stream, retrying. %s"
+                % (errMsg)
+            )
+            putRecordsToKinesisStream(
+                streamName, failedRecords, client, attemptsMade + 1, maxAttempts
+            )
+        else:
+            raise RuntimeError(
+                "Could not put records after %s attempts. %s"
+                % (str(maxAttempts), errMsg)
+            )
+
+
+def createReingestionRecord(isSas, originalRecord):
+    if isSas:
+        return {
+            "data": base64.b64decode(originalRecord["data"]),
+            "partitionKey": originalRecord["kinesisRecordMetadata"]["partitionKey"],
+        }
+    else:
+        return {"data": base64.b64decode(originalRecord["data"])}
+
+
+def getReingestionRecord(isSas, reIngestionRecord):
```

Comment:
Probably best to stick to snake case instead of switching
```suggestion
def get_reingestion_record(is_sas, re_ingestion_record):
```

Reply:
I guess if it's just copy/pasted from somewhere it's not a big deal. Kind of weird that they did mixed in the same file though (see `lambda_handler` below)

Reply:
@nbragdon That was bothering me too. I just left the original base code from Amazon alone, but might refactor it later :-)

NOTE:  I am needing to update the lambda code related to some nested JSON parsing issues with the Glue crawler. Should have that pushed up soon. Put the PR on pause until then for review.

Reply:
Put this PR back up for review via label. Found some crawler settings that resolved the nested JSON issue vs. handling via the lambda. So this code looks to work as is.